rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate protoc --proto_path=model/http-security --gogo_out=model/http-security model/http_security/httpsecurity.proto
+
+package security
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/mux"
+	"github.com/ligato/cn-infra/logging"
+	httpsecurity "github.com/ligato/cn-infra/rpc/rest/security/model/http-security"
+	"github.com/pkg/errors"
+	"github.com/unrolled/render"
+	"golang.org/x/crypto/bcrypt"
+)
+
+const (
+	// Helps to obtain authorization header matching the field in a request
+	authHeaderStr = "authorization"
+	// Used to sign the token // todo maybe use something more secure
+	secret = "secret"
+	// Admin constant, used to define admin security group and user
+	admin = "admin"
+)
+
+const (
+	// URL for login. Successful login returns token. Re-login invalidates old token and returns a new one.
+	login = "/login"
+	// URL key for logout, invalidates current token.
+	logout = "/logout"
+)
+
+// AuthenticatorAPI provides methods for handling permissions
+type AuthenticatorAPI interface {
+	// AddPermissionGroup adds new permission group. PG is defined by name and a set of URL keys. User with
+	// permission group enabled has access to that set of keys. PGs with duplicated names are skipped.
+	AddPermissionGroup(group ...*httpsecurity.PermissionGroup)
+
+	// Validate serves as middleware used while registering new HTTP handler. For every request, token
+	// and permission group is validated.
+	Validate(provider http.HandlerFunc) http.HandlerFunc
+}
+
+// Authenticator keeps information about users, permission groups and tokens and processes it
+type authenticator struct {
+	log logging.Logger
+
+	// Router instance automatically registers login/logout REST API handlers if authentication is enabled
+	router    *mux.Router
+	formatter *render.Render
+
+	// User database keeps all known users with permissions and hashed password. Users are loaded from
+	// HTTP config file // todo add other options
+	userDb []*httpsecurity.User
```

Comment:
Users and permissions should not be stored as fields directly in `authenticator`. Authenticator should access them through interface, so we can use other implementations with different storage methods. 

Create interface for retrieving these, something like:

```go
type AuthStore interface {
    GetUser(name string) (*User, error)
    GetAllUsers() ([]*User, error)
    AddUser(user *User) error
    // permissions, etc...
}
```

and then create default implementation which will store them in memory, similarity like you did here and make it possible to set different storage through authenticator constructor.